tools/codegen/api/types.py

```diff
@@ -26,13 +26,58 @@
 # you need trnsnlations that know about these types, beef up this data
 # structure.
 
-@dataclass(frozen=True)
+
+# The set of all non-templated, valid, fully-qualified names of C++ types that are used in the codegen.
+# Templated types get their own dataclass, mainly to make namespace parsing easier.
+BaseCppType = Enum('BaseCppType', (
```

Comment:
Use of an Enum here seems a bit questionable, because you're using enum values that have double colons in their names, guaranteeing that you cannot reference enum values as BaseCppType.std::string, for example

Reply:
A more common way to model this sort of situation is to define some sort of structure that records the entity name and namespace, and then define a bunch of singleton objects for each entity variant.

```
@dataclass(frozen=True)
class BaseCppType:
  ns: Optional[str]
  name: str

Int64 = BaseCppType(None, "int64_t")
Tensor = BaseCppType("at", "Tensor")
```

You stop being able to pattern match over BaseCppType but you never make use of this capability anyway.

Reply:
Using an actual class instead of an enum to capture the namespace/name pair sounds good, and I like the idea of a set of singletons that are "valid" instances of the type.